### PR TITLE
[Feature] Added pokemon nicknames/renaming

### PR DESCRIPTION
**Pull request by sirzento**
## What are the changes?

You can now rename any pokemon in your party.

## Why am I doing these changes?

This feature exist in the offical pokemon games and it was also requested a bunch of times.

## What did change?

I added a new ui that opens if you choose the rename option in the party menu. This Ui will set the pokemon `nickname`.

`nickname` is a new property of the `Pokemon` class so in theory even enemy trainer pokemon could have nicknames.

To display the nickname on all messages, most parameters of the localization functions got replaced from `pokemon.name` to `pokemon.getNameToRender()` to eighter get the nickname of that pokemon if set or the localized pokemon name.

### Renaming

https://github.com/pagefaultgames/pokerogue/assets/16718523/4886b7e6-8df4-461f-85fc-b1cb5db56368

### After evolution

https://github.com/pagefaultgames/pokerogue/assets/16718523/8e42575f-1c93-4ea2-9dae-85bd211ef63b

### Reloading

https://github.com/pagefaultgames/pokerogue/assets/16718523/d97b3fcc-f548-4829-aaa4-e4f1c24da164

### Reset back to name

https://github.com/pagefaultgames/pokerogue/assets/16718523/5546299a-f8fb-42e4-a9a3-8b0d69d16380

### Fusing
The nickname of the first pokemon gets taken. If a nicknamed fused pokemon unfuses, the nickname will stay.

## How to test the changes?

Go into the game and just rename a pokemon.

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Comment by CodeTappert** [comment]
Not a fan on how the "OK" button looks. Otherwise super cool feature!

**Comment by sirzento** [comment]
> Not a fan on how the "OK" button looks. Otherwise super cool feature!

@CodeTappert  Oh right. The first video uses an old version of the popup. The new version is shown in the second video. I will take a new one and replace it

**Comment by CodeTappert** [comment]
Hmm the buttons still could use more padding i would say. So the border doesnt overlap with the text

**Comment by sirzento** [comment]
I guess yes. The login screen could need the same change haha 

grafik

File: src/system/game-data.ts [hunk 1]
```diff
@@ -1151,6 +1151,7 @@ export class GameData {
 
   saveAll(scene: BattleScene, skipVerification: boolean = false, sync: boolean = false, useCachedSession: boolean = false, useCachedSystem: boolean = false): Promise<boolean> {
     return new Promise<boolean>(resolve => {
+
```
Comment (XGSleepWalker) [inline src/system/game-data.ts:1154]: Delete these new lines

File: src/system/game-data.ts [hunk 2]
```diff
@@ -1172,6 +1173,7 @@ export class GameData {
 
         localStorage.setItem(`data_${loggedInUser.username}`, encrypt(JSON.stringify(systemData, (k: any, v: any) => typeof v === "bigint" ? v <= maxIntAttrValue ? Number(v) : v.toString() : v), bypassLogin));
 
+
```
Comment (XGSleepWalker) [inline src/system/game-data.ts:1176]: Delete these new lines

**Review by f-fsantos — DISMISSED** [review]

File: src/locales/fr/menu.ts [hunk 3]
```diff
@@ -49,4 +49,7 @@ export const menu: SimpleTranslationEntries = {
   "disclaimer": "AVERTISSEMENT",
   "disclaimerDescription": "Ce jeu n’est pas un produit fini et peut contenir des problèmes de jouabilité, dont de possibles pertes de sauvegardes,\ndes modifications sans avertissement et pourrait ou non encore être mis à jour ou terminé.",
   "choosePokemon": "Sélectionnez un Pokémon.",
+  "renamePokemon": "Rename Pokémon",
+  "rename": "Rename",
+  "nickname": "Nickname",
```
Comment (Adri1) [inline src/locales/fr/menu.ts:54]:
```suggestion
  "renamePokemon": "Renommer Pokémon",
  "rename": "Renommer",
  "nickname": "Surnom",
```

**Review by Adri1 — COMMENTED** [review]
French translation

File: src/locales/zh_CN/menu.ts [hunk 4]
```diff
@@ -54,4 +54,7 @@ export const menu: SimpleTranslationEntries = {
   "disclaimer": "免责声明",
   "disclaimerDescription": "这个游戏尚未完成; 可能存在游戏性问题（包括潜在的丢档风险）、\n 不经通知的调整、 未来可能会更新或完成更多内容",
   "choosePokemon": "选择一只宝可梦。",
+  "renamePokemon": "Rename Pokémon",
+  "rename": "Rename",
+  "nickname": "Nickname",
```
Comment (40chyan) [inline src/locales/zh_CN/menu.ts:59]:
```suggestion
  "renamePokemon": "给宝可梦起名",
  "rename": "起名",
  "nickname": "昵称",
```

Comment (40chyan) [reply src/locales/zh_CN/menu.ts:59]: very cool enhancement to the game

File: src/locales/pt_BR/menu.ts [hunk 5]
```diff
@@ -54,4 +54,7 @@ export const menu: SimpleTranslationEntries = {
   "disclaimer": "AVISO",
   "disclaimerDescription": "Este jogo é um produto inacabado; ele pode ter problemas de jogabilidade (incluindo possíveis\n perdas de dados salvos), sofrer alterações sem aviso prévio e pode ou não ser atualizado ou concluído.",
   "choosePokemon": "Escolha um Pokémon.",
+  "renamePokemon": "Rename Pokémon",
+  "rename": "Rename",
+  "nickname": "Nickname",
```
Comment (josericardo-fo) [inline src/locales/pt_BR/menu.ts:59]:
```suggestion
  "renamePokemon": "Renomear Pokémon",
  "rename": "Renomear",
  "nickname": "Apelido",
```

File: src/locales/ko/menu.ts [hunk 6]
```diff
@@ -54,4 +54,7 @@ export const menu: SimpleTranslationEntries = {
   "disclaimer": "면책 조항",
   "disclaimerDescription": "이 게임은 완전히 개발되지 않았습니다- (세이브 데이터 소실을 포함) 플레이에 지장을 주는 문제가 생길 수 있으며,\n공지 없이 업데이트가 진행 혹은 중지될 수 있습니다.",
   "choosePokemon": "포켓몬을 선택하세요.",
+  "renamePokemon": "Rename Pokémon",
+  "rename": "Rename",
+  "nickname": "Nickname",
```
Comment (EnochG1) [inline src/locales/ko/menu.ts:59]:
```suggestion
  "renamePokemon": "포켓몬의 닉네임은?",
  "rename": "닉네임 바꾸기",
  "nickname": "닉네임",
```

**Review by EnochG1 — COMMENTED** [review]
IMO, we should localize messages also. like below
```
`It doesn\'t affect ${pokemon.getNameToRender()}!`;
getPokemonMessage(pokemon, ` stole\n${defender.getNameToRender()}'s ${stolenItem.type.name}!`)
getPokemonMessage(target, ` drank down all the\nmatcha that ${pokemon.getNameToRender()} made!`), true, !this.showAnim)
getPokemonMessage(pokemon, " frisked " + opponent.getNameToRender() + "'s " + opponent.getAbility().name + "!")
`Pointed stones dug into\n${pokemon.getNameToRender()}!`
`The opposing ${pokemon.getNameToRender()} was caught in a sticky web!`
```

It should be changed like 
```
i18next.t("abilityTriggers:windPowerCharged", { pokemonName: pokemon.getNameToRender(), moveName: this.getMoveName() })
```

**Comment by sirzento** [comment]
> IMO, we should localize messages also. like below
> 
> ```
> `It doesn\'t affect ${pokemon.getNameToRender()}!`;
> getPokemonMessage(pokemon, ` stole\n${defender.getNameToRender()}'s ${stolenItem.type.name}!`)
> getPokemonMessage(target, ` drank down all the\nmatcha that ${pokemon.getNameToRender()} made!`), true, !this.showAnim)
> getPokemonMessage(pokemon, " frisked " + opponent.getNameToRender() + "'s " + opponent.getAbility().name + "!")
> `Pointed stones dug into\n${pokemon.getNameToRender()}!`
> `The opposing ${pokemon.getNameToRender()} was caught in a sticky web!`
> ```
> 
> It should be changed like
> 
> ```
> i18next.t("abilityTriggers:windPowerCharged", { pokemonName: pokemon.getNameToRender(), moveName: this.getMoveName() })
> ```

I just changed it one to one. I thought that's something for a different PR to change all the hardcoded strings to localization.

**Comment by torranx** [comment]
im curious, wont a getter function work instead of manually creating `getNameToRender()` and changing all instances of `.name`

**Comment by RedstonewolfX** [comment]
I used this PR’s files in an offline/modded copy, and the game crashed when I hatched an egg because it had no nickname to copy (I may have copied it incorrectly though).

Also, there’s a generate name property of Pokemon.
When I tried to implement nicknames myself, I made it output the nickname if there was one, and set the name as normal if there wasn’t. You can make it call this function whenever the name is changed

**Comment by sirzento** [comment]
> I used this PR’s files in an offline/modded copy, and the game crashed when I hatched an egg because it had no nickname to copy (I may have copied the files incorrectly though).

Hatching eggs does work for me. Might be that you did something wrong or you did something that I didn't test but since getting eggs from the gacha and hatching them works I don't know what else you might did^^


> Also, there’s a generate name property of Pokemon.

This is used to generate the name for fused pokemon and also changing this function would lead to the same problems as I have described on my last comment.


> When I tried to implement nicknames myself, I made it output the nickname if there was one, and set the name as normal if there wasn’t. You can make it call this function whenever the name is changed

We don't want to override the name property of `Pokemon` so I really don't think it would be much to use doing it in that function.

**Comment by RedstonewolfX** [comment]
This is a very cool feature, though, and I hope it gets added

**Comment by sirzento** [comment]
> so we want to replace "pokemon.name" everywhere with getPokemonWithAffix(pokemon).
> 
> This is not compatible with your change. Any ideas?

@CodeTappert  I think it is since I also changed the name inside the getPokemonWithAffix function. So it should work I guess.

File: src/data/ability.ts [hunk 7]
```diff
@@ -573,7 +573,7 @@ export class MoveImmunityAbAttr extends PreDefendAbAttr {
   }
 
   getTriggerMessage(pokemon: Pokemon, abilityName: string, ...args: any[]): string {
-    return `It doesn\'t affect ${pokemon.name}!`;
+    return `It doesn\'t affect ${pokemon.getNameToRender()}!`;
```
Comment (CodeTappert) [inline src/data/ability.ts:576]: use getPokemonNameWithAffix everywhere

**Comment by CodeTappert** [comment]
> > so we want to replace "pokemon.name" everywhere with getPokemonWithAffix(pokemon).
> > This is not compatible with your change. Any ideas?
> 
> @CodeTappert I think it is since I also changed the name inside the getPokemonWithAffix function. So it should work I guess.

THen use getPokemonWithAffix  everywhere

**Comment by sirzento** [comment]
> > > so we want to replace "pokemon.name" everywhere with getPokemonWithAffix(pokemon).
> > > This is not compatible with your change. Any ideas?
> > 
> > 
> > @CodeTappert I think it is since I also changed the name inside the getPokemonWithAffix function. So it should work I guess.
> 
> THen use getPokemonWithAffix everywhere

Done. Some localizations needs do be changed now but I don't know if this all is still inside the scope of this PR. Like the battle.ts localization: `"singleWildAppeared": "A wild {{pokemonName}} appeared!",` will now print "A wild Mewtwo (wild) appeared!" cause of the affix. I can change it in the english localization but I have no clue about the other languages.

**Comment by CodeTappert** [comment]
> > > > so we want to replace "pokemon.name" everywhere with getPokemonWithAffix(pokemon).
> > > > This is not compatible with your change. Any ideas?
> > > 
> > > 
> > > @CodeTappert I think it is since I also changed the name inside the getPokemonWithAffix function. So it should work I guess.
> > 
> > THen use getPokemonWithAffix everywhere
> 
> Done. Some localizations needs do be changed now but I don't know if this all is still inside the scope of this PR. Like the battle.ts localization: `"singleWildAppeared": "A wild {{pokemonName}} appeared!",` will now print "A wild Mewtwo (wild) appeared!" cause of the affix. I can change it in the english localization but I have no clue about the other languages.

Please list all localization keys that needs this Change

**Comment by sirzento** [comment]
> > > > > so we want to replace "pokemon.name" everywhere with getPokemonWithAffix(pokemon).
> > > > > This is not compatible with your change. Any ideas?
> > > > 
> > > > 
> > > > @CodeTappert I think it is since I also changed the name inside the getPokemonWithAffix function. So it should work I guess.
> > > 
> > > 
> > > THen use getPokemonWithAffix everywhere
> > 
> > 
> > Done. Some localizations needs do be changed now but I don't know if this all is still inside the scope of this PR. Like the battle.ts localization: `"singleWildAppeared": "A wild {{pokemonName}} appeared!",` will now print "A wild Mewtwo (wild) appeared!" cause of the affix. I can change it in the english localization but I have no clue about the other languages.
> 
> Please list all localization keys that needs this Change

All localizations that have "wild" inside them. I currently only know about "singleWildAppeared" and "multiWildAppeared" and I don't think there are more like that.

**Comment by CodeTappert** [comment]
Then dont do it for those two.

**Comment by sirzento** [comment]
> Then dont do it for those two.

Alright. Chnaged those two back

**Comment by CodeTappert** [comment]
image

**Comment by CodeTappert** [comment]
https://github.com/user-attachments/assets/5fc88e9a-7477-407c-b24a-96e8f2c88f4b

**Comment by CodeTappert** [comment]
Setting name -> F5 -> Name gone

**Comment by CodeTappert** [comment]
Setting a weird name like:
image

Leads to 

image


So you need to disallow some characters or fix that bug

**Comment by sirzento** [comment]
> image

Ok fixed. Yeah I didn't checked that one.

**Comment by sirzento** [comment]
> Setting name -> F5 -> Name gone

Yes because the game doesn't get saved if you press F5. Should the game be saved after a rename?

**Comment by flx-sta** [comment]
> Setting a weird name like: image
> 
> Leads to
> 
> image
> 
> So you need to disallow some characters or fix that bug

I think it would be cool to keep such weird names allowed
This shouldn't be an issue and we need to support non latin1 characters anyway for chinese/japanese and stuff

You just gotta [escape them](https://stackoverflow.com/a/56647993/4622620) before base64 encoding them

**Comment by sirzento** [comment]
> > Setting a weird name like: [image](https://private-user-images.githubusercontent.com/38758606/349591321-5eb70a27-0736-4b4d-bdef-b41b31686242.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEyMzcwNjQsIm5iZiI6MTcyMTIzNjc2NCwicGF0aCI6Ii8zODc1ODYwNi8zNDk1OTEzMjEtNWViNzBhMjctMDczNi00YjRkLWJkZWYtYjQxYjMxNjg2MjQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE3VDE3MTkyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxMjEyN2M5NzU3NGY3ZWE3ZmQxNzliODc3ZDYxNjU3YWQxOTY3NTNiZDY2ZTQ5MjIzOTVlYTY2MTA2NDc4MDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Swmzu5LuwgOe3EaEkk-vnt-X5ipd4iMJNjMt3uFgIi4)
> > Leads to
> > [image](https://private-user-images.githubusercontent.com/38758606/349590784-679633e5-ec5c-4890-af9c-8862ee54b1ed.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEyMzcwNjQsIm5iZiI6MTcyMTIzNjc2NCwicGF0aCI6Ii8zODc1ODYwNi8zNDk1OTA3ODQtNjc5NjMzZTUtZWM1Yy00ODkwLWFmOWMtODg2MmVlNTRiMWVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE3VDE3MTkyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ0ODU3OGMyZTczZjE4ZGYwOGU4NDM0ZTI5YWQwZmRkNGU3MzQ1ZTljNDY4ZDFhZTllZjFlNmQyZDFmYmEwZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Vo6IXPfJA-45ILEj7YjvXVsDSht--W1Re1-7-G8Amck)
> > So you need to disallow some characters or fix that bug
> 
> I think it would be cool to keep such weird names allowed This shouldn't be an issue and we need to support non latin1 characters anyway for chinese/japanese and stuff
> 
> You just gotta [escape them](https://stackoverflow.com/a/56647993/4622620) before base64 encoding them

It's so cursed haha
grafik

File: src/field/pokemon.ts [hunk 8]
```diff
@@ -226,6 +229,19 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
 
+
+  getNameToRender() {
+    try {
+      if (this.nickname) {
+        return decodeURIComponent(escape(atob(this.nickname)));
+      }
+      return this.name;
+    } catch (err) {
+      console.error(err);
```
Comment (flx-sta) [inline src/field/pokemon.ts:240]: Please add some context to the error
```suggestion
      console.error(`Failed to decode nickname for ${this.name}`, err);
```

**Review by flx-sta — CHANGES_REQUESTED** [review]

**Review by flx-sta — APPROVED** [review]

**Review by CodeTappert — APPROVED** [review]

**Comment by CodeTappert** [comment]
image

I mean this time the game at least doesnt hang. But maybe have a look at it

**Comment by CodeTappert** [comment]
Otherwise this fully works.

**Comment by flx-sta** [comment]
Maybe we also wanna change this to be a `console.warn()` since it's not crucial, just nice to have

**Comment by sirzento** [comment]
> image
> 
> I mean this time the game at least doesnt hang. But maybe have a look at it
> 
> (i named the pokemon '✌✌✌✌✌✌' )

Mh it does work on my end when I copy those emojis 
grafik

**Comment by flx-sta** [comment]
I can confirm that it works fine (english, legacy theme).
I even tried importing/exporting sessions =>. no problem

**Comment by CodeTappert** [comment]
> > image
> > I mean this time the game at least doesnt hang. But maybe have a look at it
> > (i named the pokemon '✌✌✌✌✌✌' )
> 
> Mh it does work on my end when I copy those emojis. Do you have the latest version? grafik

As i said. it WORKS. But still produces those messages in the console on setting that name. I can play without problems

**Comment by flx-sta** [comment]
@CodeTappert I don't see this console log issue

@sirzento But I found an issue: When writing the name it seems to notice some input still as ingame input. It might event get stuck in a `press-down` loop.
that is a big problem. (You can see that slightly in the video. When it constantly moves down in the shop. that wasn't me pressing any button doing that). YOu have to make sure to suppress that these inputs are propagated down

https://github.com/user-attachments/assets/45b1a1b4-759a-496d-b784-09eac0ab9975

**Comment by sirzento** [comment]
@CodeTappert Yeah I actually ment that I also don't get those console outputs.

@flx-sta Oh okey. I have no clue on how to fix or prevent this. So this happens if you press buttons like left/down/right/up while in the text input field? I never got that behavior before.

**Comment by flx-sta** [comment]
Yeah. It could be due to that specific Mac OS emoji menu thiough. So extremely specific

File: src/phases.ts [hunk 9]
```diff
@@ -1438,7 +1438,7 @@ export class SummonPhase extends PartyMemberPokemonPhase {
       this.scene.time.delayedCall(750, () => this.summon());
     } else {
       const trainerName = this.scene.currentBattle.trainer.getName(!(this.fieldIndex % 2) ? TrainerSlot.TRAINER : TrainerSlot.TRAINER_PARTNER);
-      const pokemonName = this.getPokemon().name;
+      const pokemonName = getPokemonNameWithAffix(this.getPokemon());
```
Comment (flx-sta) [inline src/phases.ts:1441]: This seems to cause wrong behavior @sirzento 
image

**Pull request merged**